### ISSUE TITLE
CyclomaticComplexity reports phantom violations on simple methods when 3+ files share a private method name

### ISSUE THREAD

**Issue by chadmandoo** [phpmd/phpmd#1286]
- PHPMD version: 2.15.0
- PHP Version: 8.5
- Installation type: composer (require-dev)
- Operating System: Linux (Ubuntu 24.04)

### Current Behavior
When PHPMD is invoked on a comma-separated list of files where 3+ files contain private methods sharing the same name, the `CyclomaticComplexity` rule's analysis state appears to leak across files: methods with real CC=2 are reported as CC=8 (or whatever value matches a structurally distinct same-named method's CC in the set).

### Expected Behavior
Each method's CyclomaticComplexity should be computed in isolation. A method with real CC=2 must never be flagged when `reportLevel=8`, regardless of what other files in the same invocation contain.

### Steps To Reproduce

Three files, each with `private function sharedHelper(string $path): void`:

`a.php` — CC ≈ 5 (iterator-based recursion):
```php
<?php
class A {
    private function sharedHelper(string $path): void {
        if (!is_dir($path)) return;
        $it = new \RecursiveIteratorIterator(
            new \RecursiveDirectoryIterator($path, \RecursiveDirectoryIterator::SKIP_DOTS),
            \RecursiveIteratorIterator::CHILD_FIRST
        );
        foreach ($it as $f) {
            if ($f->isDir()) { rmdir($f->getPathname()); }
            else { unlink($f->getPathname()); }
        }
        rmdir($path);
    }
}
```

`b.php` and `c.php` — CC = 2 (single guard):
```php
<?php
class B { // and C, identically
    private function sharedHelper(string $path): void {
        if (is_dir($path)) {
            // delete via subprocess
        }
    }
}
```

`ruleset.xml`:
```xml
<?xml version="1.0"?>
<ruleset name="repro">
  <rule ref="rulesets/codesize.xml/CyclomaticComplexity">
    <properties><property name="reportLevel" value="8"/></properties>
  </rule>
</ruleset>
```

Invocation matrix:

| Command | Result |
| ------- | ------ |
| `phpmd a.php text ruleset.xml` | exit 0 (clean) |
| `phpmd b.php text ruleset.xml` | exit 0 (clean) |
| `phpmd c.php text ruleset.xml` | exit 0 (clean) |
| `phpmd a.php,b.php text ruleset.xml` | exit 0 (clean) |
| `phpmd a.php,b.php,c.php text ruleset.xml` | **exit 2** — `b.php`/`c.php` flagged with phantom CC=8; real CC=2 method in `a.php` is NOT flagged |

The leak appears to require **3+ files** with the shared name and **at least one body** with CC ≈ 5.

### Notes
- Pre-existing closed issue #1191 covers a separate performance regression on multi-file invocation; this is a correctness bug.
- Workaround in our project: rename the helpers per-file to break the trigger pattern.

## Checks before submitting
* [x] No duplicate issue/PR found
* [x] Steps to reproduce included
* [x] Code example provided
* [x] Single-bug scope
* [x] Descriptive title

**Comment by AJenbo** [comment]
2.x does not have PHP 8.5 support, could you please test this with 3.x-dev to verify if it's still an issue.

**Comment by chadmandoo** [comment]
Thanks @AJenbo — tested against 3.x-dev (commit `35140602`) in an isolated repro env. **The bug does not reproduce on 3.x-dev.**

### Findings

PHPMD 3.x removed the legacy comma-separated multi-path syntax entirely — it's now Symfony Console with `analyze` as the default command and space-separated positional `<paths>...`. The 2.x repro recipe (`phpmd a.php,b.php,c.php text ruleset.xml`) errors with `Engine.php line 222: The given file "a.php,b.php,c.php" does not exist.` rather than running.

With the new 3.x CLI shape, the multi-file matrix is clean across the board:

| 3.x invocation | Exit code | Notes |
|---|---|---|
| `phpmd analyze --format=text --ruleset=ruleset.xml src/A.php` | 0 | clean |
| `phpmd analyze --format=text --ruleset=ruleset.xml src/B.php` | 0 | clean |
| `phpmd analyze --format=text --ruleset=ruleset.xml src/C.php` | 0 | clean |
| `phpmd analyze --format=text --ruleset=ruleset.xml src/A.php src/B.php` | 0 | clean |
| `phpmd analyze --format=text --ruleset=ruleset.xml src/A.php src/B.php src/C.php` | 0 | clean — **the original 3-file aggregation case** |
| `phpmd analyze --format=text --ruleset=ruleset.xml src/` | 0 | clean (directory walk) |
| `phpmd analyze --format=text --ruleset=ruleset.xml src/A.php,src/B.php,src/C.php` | 1 | "file does not exist" — legacy CSV path removed |

Repro fixture: same three files (`A.php` with the iterator-based CC≈5 helper, `B.php` and `C.php` with the simple `if (is_dir)` CC=2 helpers) and the same `ruleset.xml` with `CyclomaticComplexity reportLevel=8`. Installed via:

```
composer require --dev phpmd/phpmd:3.x-dev
```

(resolved to `phpmd/phpmd 3.x-dev 35140602568b6068884b28e04b3dcf1cde9d5ea1`)

### Verdict

Closable as fixed-in-3.x — either the Symfony Console refactor incidentally cleared the CC analysis-state leak, or the leak was always invocation-shape-dependent and the new shape happens to avoid it. From a 2.x consumer's perspective, the practical fix is "upgrade to 3.x" once stable, with the additional migration note that the CSV multi-path syntax must be replaced with space-separated positional arguments.

Happy for this to close once you confirm. Thank you for the prompt response and for the heads-up about 2.x's PHP 8.5 gap.

**Comment by AJenbo** [comment]
Ok we will upgrade the demo project to use 3.x once stable thank you for the detailed anaplasis, but I'm going to go on a limb and say that the parser bug was not fixed by the switch to Symfony Console 🤖
